### DISCUSSION TITLE
tree-sitter doesn't support look-ahead, how can I implement it?

### DISCUSSION THREAD

**Discussion by shi-yan** [tree-sitter/tree-sitter#1005] [Q&A]
I'm trying to parse the following syntax:

```
begin
/[^\n]+\n/
/[^\n]+\n/
/[^\n]+\n/
...
end
```

basically, between "begin" and "end", there are multiple lines of content.

I'm doing this currently

```
content: $ => seq($.begin, '\n', repeat(/[^\n]+\n/), $.end),
```

this won't work, because

"end" will match /[^\n]+\n/, which confuses the parser.

so I want to add a look-ahead:

```
content: $ => seq($.begin, '\n', repeat(/(?!'end')[^\n]+\n/),$.end),
```

and then I got

```
error: look-around, including look-ahead and look-behind, is not supported
```

How can I implement this then?

**Comment by maxbrunsfeld** [comment]
I converted this into a discussion thread, since we'd prefer to use the issues list only for bug reports and feature requests.

**Reply by shi-yan** [reply]
sorry, I did looked at the document before posing this issue. I didn't find where I should post discussions. 

I didn't realize there is a discussion board here. Thanks.

**Reply by shi-yan** [reply]
I still can't find a good solution to my issue.

the doc does have a section listing some solutions:

https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens

right now, only "Match Specificity" works.


I'm trying to parse plantuml https://plantuml.com/sequence-diagram#012d3e8694a98cc1

if I hard code the end token as "end note", it will work.

but if I use regexp /"end"\s+"note"/, it will fail. So there can only be one space between "end" and "note".

Could we explicitly support look-ahead, or rule orders (I tried moving the preferred rule above, doesn't seem to work, I hope we can specify order explicitly, the current prec thing doesn't work in my situation.)

**Reply by shi-yan** [reply]
does it make sense to introduce a new function "not"

not(rule1, rule2)

accept rule2 only when it can't be matched by rule1

**Comment by ahlinc** [comment]
@shi-yan If it's possible, It would be helpful if you would create a repo according to [the tree-sitter repo setup guide](https://tree-sitter.github.io/tree-sitter/creating-parsers#project-setup) and give a link to it, this would be very informative in understanding a context and help to understand how you configured your grammar.
